### PR TITLE
CURA-5389 Check and confirmation before Remove package

### PR DESCRIPTION
**Pull request by jackha**
Before removing a package, the contents are inspected to see if anything is in use. If there is, you'll get a popup to confirm that the affected machines are then set back to their default material / profile before uninstalling.

### CODE REVIEW

File: plugins/Toolbox/resources/qml/ToolboxConfirmUninstallResetDialog.qml [hunk 1]
```diff
@@ -19,7 +19,7 @@ UM.Dialog
 
     title: catalog.i18nc("@title:window", "Confirm uninstall ") + toolbox.pluginToUninstall
     width: 450 * screenScaleFactor
-    height: (30 + dialogText.height) * screenScaleFactor
+    height: (100 + dialogText.height) * screenScaleFactor
```
Comment (fieldOfView) [inline plugins/Toolbox/resources/qml/ToolboxConfirmUninstallResetDialog.qml]: Should be ```100 * screenScaleFactor + dialogText.height```
dialogText.height is already DPI-aware.

Comment (jackha) [reply plugins/Toolbox/resources/qml/ToolboxConfirmUninstallResetDialog.qml]: Ahhh thanks ^^

File: cura/CuraPackageManager.py [hunk 2]
```diff
@@ -15,3 +17,23 @@ def initialize(self):
         self._installation_dirs_dict["qualities"] = Resources.getStoragePath(CuraApplication.ResourceTypes.QualityInstanceContainer)
 
         super().initialize()
+
+    ##  Returns a list of where the package is used
+    #   empty if it is never used.
+    #   It loops through all the package contents and see if some of the ids are used.
+    #   The list consists of 3-tuples: (global_stack, extruder_nr, container_id)
+    def getMachinesUsingPackage(self, package_id: str):
```
Comment (ianpaschal) [inline cura/CuraPackageManager.py]: Needs return typing.

File: cura/Machines/MaterialGroup.py [hunk 3]
```diff
@@ -18,11 +21,11 @@
...
     __slots__ = ("name", "is_read_only", "root_material_node", "derived_material_node_list")
 
-    def __init__(self, name: str, root_material_node: MaterialNode) -> None:
+    def __init__(self, name: str, root_material_node: "MaterialNode") -> None:
         self.name = name
         self.is_read_only = False
-        self.root_material_node = root_material_node # type: MaterialNode
-        self.derived_material_node_list = [] #type: List[MaterialNode]
+        self.root_material_node = root_material_node  # type: MaterialNode
+        self.derived_material_node_list = []  # type: List[MaterialNode]
```
Comment (ianpaschal) [inline cura/Machines/MaterialGroup.py]: Don't need double spaces here before comment.

File: cura/Machines/QualityManager.py [hunk 4]
```diff
@@ -340,6 +340,13 @@ def getQualityGroupsForMachineDefinition(self, machine: "GlobalStack") -> dict:
 
         return quality_group_dict
 
+    def getDefaultQualityType(self, machine: "GlobalStack") -> Optional[QualityGroup]:
+        preferred_quality_type = machine.definition.getMetaDataEntry("preferred_quality_type")
+        quality_group_dict = self.getQualityGroups(machine)
+        quality_group = quality_group_dict.get(preferred_quality_type)
```
Comment (ianpaschal) [inline cura/Machines/QualityManager.py]: Not sure if tests will complain that if `quality_group_dict` or `preferred_quality_type` are None, this line will not work.

File: plugins/Toolbox/resources/qml/ToolboxConfirmUninstallResetDialog.qml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    Column
+    {
+        UM.I18nCatalog { id: catalog; name: "cura" }
+
+        anchors.fill: parent
+        anchors.leftMargin: Math.round(20 * screenScaleFactor)
+        anchors.rightMargin: Math.round(20 * screenScaleFactor)
+        anchors.topMargin: Math.round(10 * screenScaleFactor)
+        anchors.bottomMargin: Math.round(10 * screenScaleFactor)
```
Comment (ianpaschal) [inline plugins/Toolbox/resources/qml/ToolboxConfirmUninstallResetDialog.qml]: Personal preference but I think its a lot cleaner when this sort of thing is written as:

```
anchors
{
    fill: parent
    leftMargin: ... etc
}
```

File: plugins/Toolbox/src/Toolbox.py [hunk 6]
```diff
@@ -305,9 +314,87 @@ def install(self, file_path: str) -> None:
...
+        machine_manager = application.getMachineManager()
+
+        for global_stack, extruder_nr, container_id in self._package_used_materials:
+            default_material_node = material_manager.getDefaultMaterial(global_stack, extruder_nr, global_stack.extruders[extruder_nr].variant.getName())
+            machine_manager.setMaterial(extruder_nr, default_material_node, global_stack = global_stack)
+        for global_stack, extruder_nr, container_id in self._package_used_qualities:
+            default_quality_group = quality_manager.getDefaultQualityType(global_stack)
+            machine_manager.setQualityGroup(default_quality_group, global_stack = global_stack)
+
+        self._markPackageMaterialsAsRemove(self._package_id_to_uninstall)
```
Comment (ianpaschal) [inline plugins/Toolbox/src/Toolbox.py]: Could have a more clear name to make it easier to know if stuff is being marked "as removed" or "as removable" or what. "remove" is kind of ambiguous.

File: plugins/UM3NetworkPrinting/LegacyUM3OutputDevice.py [hunk 7]
```diff
@@ -458,15 +458,9 @@ def _onCheckAuthenticationFinished(self, reply):
     def _saveAuthentication(self):
         global_container_stack = CuraApplication.getInstance().getGlobalContainerStack()
         if global_container_stack:
-            if "network_authentication_key" in global_container_stack.getMetaData():
-                global_container_stack.setMetaDataEntry("network_authentication_key", self._authentication_key)
-            else:
-                global_container_stack.addMetaDataEntry("network_authentication_key", self._authentication_key)
+            global_container_stack.setMetaDataEntry("network_authentication_key", self._authentication_key)
```
Comment (ianpaschal) [inline plugins/UM3NetworkPrinting/LegacyUM3OutputDevice.py]: Should validate `self._authentication_key`

File: plugins/UM3NetworkPrinting/LegacyUM3OutputDevice.py [hunk 8]
```diff
@@ -458,15 +458,9 @@ def _onCheckAuthenticationFinished(self, reply):
...
 
-            if "network_authentication_id" in global_container_stack.getMetaData():
-                global_container_stack.setMetaDataEntry("network_authentication_id", self._authentication_id)
-            else:
-                global_container_stack.addMetaDataEntry("network_authentication_id", self._authentication_id)
+            global_container_stack.setMetaDataEntry("network_authentication_id", self._authentication_id)
```
Comment (ianpaschal) [inline plugins/UM3NetworkPrinting/LegacyUM3OutputDevice.py]: Should validate `self._authentication_id`

**Comment by ianpaschal** [comment]
Also, from CI:

```
plugins\Toolbox\src\Toolbox.py:333: error: Item "None" of "Optional[Any]" has no attribute "show"
```

**Pull request merged**